staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -169,10 +172,16 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 	glog.V(3).Infof("Listing and watching %v from %s", r.expectedType, r.name)
 	var resourceVersion string
 
-	// Explicitly set "0" as resource version - it's fine for the List()
-	// to be served from cache and potentially be delayed relative to
-	// etcd contents. Reflector framework will catch up via Watch() eventually.
-	options := metav1.ListOptions{ResourceVersion: "0"}
+	// Explicitly set resource version to have it list from cache for
+	// performance reasons.
+	// It's fine for the returned state to be stale (we will catch up via
+	// Watch() eventually), but can't set "0" to avoid going back in time
+	// if we hit apiserver that is significantly delayed compared to the
+	// state we already had.
+	// TODO: There is still a potential to go back in time after component
+	// restart when we set ResourceVersion: "0". For more details see:
+	// https://github.com/kubernetes/kubernetes/issues/59848
+	options := metav1.ListOptions{ResourceVersion: r.LastSyncResourceVersion()}
```

Comment:
what does the apiserver do if you ask to List with a resourceVersion that has been compacted and/or is no longer in the watch cache event buffer? does it respond with a `410 Gone` error? if so, does anything on the client-side handle 410 errors on initial List correctly?

Reply:
this was discussed but never resolved in https://github.com/kubernetes/kubernetes/issues/37473#issuecomment-265382070. I don't think resourceVersions other than 0 are thoroughly tested yet.

Reply:
looks like behavior of the API for List with a non-0 resourceVersion differs when the watch cache is enabled and when it is disabled.

with the watch cache enabled, the API has the "at least as new as" behavior

with the watch cache disabled:
* successful responses return the List as it existed at that resourceVersion (resources are returned as they existed at that version, newer resources are not included)
* if the response has been compacted, you get 410 errors:
```
$ curl -k https://localhost:6443/api/v1/configmaps?resourceVersion=1
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "The resourceVersion for the provided list is too old.",
  "reason": "Expired",
  "code": 410
}
```

we cannot switch clients to using an API with behavior that unstable

Reply:
> successful responses return the List as it existed at that resourceVersion (resources are returned as they existed at that version, newer resources are not included)

I think that is consistent with the "at least as fresh as RV" semantic.

> if the requested version has already been compacted, you get 410 errors:

This is what is causing a problem.

Thinking more about it - we can easily change watchcache, to return 410 if such version doesn't exist.
However, that doesn't make much sense with "at least as fresh as RV" semantic.
And the semantic of "exactly from a given RV" is not possible to implement efficiently in the way how watchcache is currently implemented.

So Jordan - do you suggest to drop this PR? Or do you have any better suggestion?

Reply:
@wojtek-t Instead of using "exactly from a given RV" semantics in the watchcache. Would it be feasible to switch to using "at least as fresh as RV" semantics for when watch cache is disabled? That can be supported by etcd easily, just perform a quorum read and verify the resulting RV is recent enough.  I don't know how we'd do that without breaking compatibility with the previous behavior for the impacted interfaces like ListerWatcher, though.

Reply:
> That can be supported by etcd easily, just perform a quorum read and verify the resulting RV is recent enough.

Yeah - that have a potential to work.

> I don't know how we'd do that without breaking compatibility with the previous behavior for the impacted interfaces like ListerWatcher, though.

Not sure I understand this - can you clarify?

@liggitt - what do you think about changing the semantic for "without cache" case to what Joe described above?
This is backward incompatible change, but as you already pointed out before the behavior isn't well defined now. So maybe making that change is actually fine?

I'm happy to work on that, but I think we need to first agree that that's the way to proceed.

Reply:
Re: backward incompatibility. The underlying issue here is that we have a single `storage.Interface` that is implemented with different RV semantics by `cacher.Cacher` (minimum RV semantics) than for etcd.etcdHelper and etcd3.store (exact RV semantics). This causes a couple problems:
 - We cannot document storage.Interface's RV semantics since they're not consistently across implementations
 - Any new storage.Interface implementations do not have

 So the option for addressing the problem:
 1. Break backward compatibility- like discussed above. The scope of impact is clients of kubeapiserver's where the watchcache is disabled and that expect "exact RV" semantics (is this a very small/empty group?)
 2. Introduce a new `MinimumResourceVersion` field to ListOptions that is used for "minimum RV" semantics, mark "ResourceVersions" as deprecated and document that it has inconsistent semantics. This may also require making
    changes to storage.Interface since it currently only supports a single ResourceVersion param for many functions.

 I'm currently in favor of #1.

+ @lavalamp for any thoughts

Reply:
Problem is that the cacher semantics are the the most directly exposed through the API to clients, and as a result, are what clients depend on the most, and would be most impacted by if changed. Pragmatically shouldn't we consider if the cacher semantics are what we want, and if so, start shifting system toward them?

+1 to considering component restart here as well.

Reply:
Instead of making a decision out of this PR, let's back out to a design proposal, I'll start getting something prepared and circulate it

Reply:
I’ll echo Jordan’s caution here.  The watch cache is an optimization that has proven very difficult to keep correct.  For a long time I’ve wondered whether we should be using the etcd proxy with modified semantics to get our performance trade offs vs continuing to double down on a different abstraction.

A design proposal is much needed.

Reply:
> The watch cache is an optimization that has proven very difficult to keep correct.

I would call it this way - we found a bunch bugs because of not good enough test coverage (and somewhat implicit instead of explicit semantics of the interface in some cases).

> For a long time I’ve wondered whether we should be using the etcd proxy 

That won't solve our issues - this won't avoid multiple deserialization of the same stuff which is why we mostly introduced the watch cache.

>  A design proposal is much needed.

+1 for that (and considering that together with restart issue). Thanks for taking that @jpbetz

Reply:
We could modify the etcd proxy to maintain a lookaside deserialized cache with the correct locking semantics (whether that’s the right thing to do is an open question, but it would replace our “fake etcd” semantics)